src/components/theme/ThemeProvider.tsx

```diff
@@ -1,106 +1,59 @@
-import React, {PropsWithChildren} from 'react';
+import React, {PropsWithChildren, useEffect, useMemo, useState} from 'react';
 
-import {ThemeContext, ThemeContextProps} from './ThemeContext';
-import {ThemeValueContext} from './ThemeValueContext';
 import {DEFAULT_THEME} from './constants';
+import {getThemeValue} from './getThemeValue';
+import {ThemeContext} from './ThemeContext';
+import {ThemeValueContext} from './ThemeValueContext';
+import type {Theme} from './types';
+import {updateBodyClassName} from './updateBodyClassName';
+import {useSystemTheme} from './useSystemTheme';
 
 interface ThemeProviderExternalProps {}
 
 interface ThemeProviderDefaultProps {
-    theme: string;
+    theme: Theme;
 }
 
 export interface ThemeProviderProps
     extends ThemeProviderExternalProps,
-        Partial<ThemeProviderDefaultProps> {}
-
-interface ThemeProviderState extends ThemeContextProps {
-    themeValue: string;
-}
-
-const getDarkMediaMatch = () => window.matchMedia('(prefers-color-scheme: dark)');
-function getSystemTheme() {
-    if (typeof window === 'object') {
-        return getDarkMediaMatch().matches ? 'dark' : 'light';
-    } else {
-        return 'light';
-    }
-}
-
-export class ThemeProvider extends React.Component<
-    ThemeProviderExternalProps & ThemeProviderDefaultProps & PropsWithChildren<{}>,
-    ThemeProviderState
-> {
-    static defaultProps: ThemeProviderDefaultProps = {
-        theme: DEFAULT_THEME,
-    };
-
-    mediaListener?: (event: MediaQueryListEvent) => void;
-
-    state: ThemeProviderState = {
-        theme: this.props.theme,
-        themeValue: this.getThemeValue(this.props.theme),
-        setTheme: (theme: string) => {
-            this.setState({theme});
-        },
-    };
-
-    componentDidMount() {
-        if (typeof window === 'object') {
-            this.mediaListener = (event) => {
-                if (this.state.theme === 'system') {
-                    const themeValue = event.matches ? 'dark' : 'light';
-                    this.setState({themeValue}, () => this.updateBodyClassName(themeValue));
-                }
-            };
-            getDarkMediaMatch().addListener(this.mediaListener);
+        Partial<ThemeProviderDefaultProps>,
+        PropsWithChildren<{}> {}
+
+export function ThemeProvider({theme: themeProp = DEFAULT_THEME, children}: ThemeProviderProps) {
+    const [theme, setTheme] = useState<Theme>(themeProp);
+    useEffect(() => {
+        setTheme(themeProp);
+    }, [themeProp]);
+
+    const [themeValue, setThemeValue] = useState(getThemeValue(theme));
+    const systemTheme = useSystemTheme();
+    useEffect(() => {
+        if (systemTheme && theme === 'system') {
+            setThemeValue(systemTheme);
+
+            return;
```

Comment:
let's simplify this:
```suggestion
    const systemTheme = useSystemTheme();
    const themeValue = theme === 'system' ? systemTheme : theme;
```